server-go/filtering/common/source_test.go

```diff
@@ -0,0 +1,108 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package common defines types common to parsing and other diagnostics.
+package common
+
+import (
+	"testing"
+)
+
+// Test the error description method.
+func TestStringSource_Description(t *testing.T) {
+	contents := "example content\nsecond line"
+	source := NewStringSource(contents, "description-test")
+	assertEquals(t, source.Content(), contents)
+	assertEquals(t, source.Description(), "description-test")
+	str2, found := source.Snippet(2)
+	assertTrue(t, found)
+	assertEquals(t, "second line", str2)
+	str1, found := source.Snippet(1)
+	assertTrue(t, found)
+	assertEquals(t, "example content", str1)
+}
+
+// Test the character offest to make sure that the offsets accurately reflect
+// the location of a character in source.
+func TestStringSource_CharacterOffset(t *testing.T) {
+	contents := "c.d &&\n\t b.c.arg(10) &&\n\t test(10)"
+	source := NewStringSource(contents, "offset-test")
+	assertHasExactly(t, []int32{7, 24, 35}, source.LineOffsets())
+	charStart, _ := source.CharacterOffset(NewLocation(1, 2))
+	charEnd, _ := source.CharacterOffset(NewLocation(3, 2))
+	assertEquals(t, "d &&\n\t b.c.arg(10) &&\n\t ", string(contents[charStart:charEnd]))
+	if _, found := source.CharacterOffset(NewLocation(4, 0)); found {
+		t.Error("Character offset was out of range of source, but still found.")
+	}
+}
+
+// Test the computation of snippets, single lines of text, from a multiline
+// source.
+func TestStringSource_SnippetMultiline(t *testing.T) {
+	source := NewStringSource("hello\nworld\nmy\nbub\n", "four-line-test")
+	str, found := source.Snippet(1)
+	assertTrue(t, found)
+	assertEquals(t, "hello", str)
+	str2, found := source.Snippet(2)
+	assertTrue(t, found)
+	assertEquals(t, "world", str2)
+	str3, found := source.Snippet(3)
+	assertTrue(t, found)
+	assertEquals(t, "my", str3)
+	str4, found := source.Snippet(4)
+	assertTrue(t, found)
+	assertEquals(t, "bub", str4)
+	str5, found := source.Snippet(5)
+	assertTrue(t, found)
+	assertEquals(t, "", str5)
+}
+
+// Test the computation of snippets from a single line source.
+func TestStringSource_SnippetSingleline(t *testing.T) {
+	source := NewStringSource("hello, world", "one-line-test")
+	str, found := source.Snippet(1)
+	assertTrue(t, found)
+	assertEquals(t, "hello, world", str)
+	str2, found := source.Snippet(2)
+	assertFalse(t, found)
+	assertEquals(t, "", str2)
+}
+
+// Helper assertions.
+func assertTrue(t *testing.T, actual bool) {
```

Comment:
Can we just inline these functions? I don't think they make the tests more readable. Alternatively can you call t.Helper() so the error messages include the actual line in the test function that causes the error?

Reply:
Yes, definitely. I'll inline them.